[feat]: brand new enums (#69)
soramitsu/scale-codec-js-library: packages/enum/CHANGELOG.md

# @scale-codec/enum

## 2.0.0

### Major Changes

> - 7e81bbc: **refactor**: introduce a complete new design of tagged unions

Reviewer: ```suggestion
- 7e81bbc: **refactor**: introduce a completely new design of tagged unions
```

```suggestion
- 7e81bbc: **refactor**: introduce a new design of tagged unions
```

  #### What is the change

  Enums are declared with `Enumerate<T>` type.

  **Enum in Rust:**

  ```rust
  enum Message {
    Ping,
    Text(String),
    Coords(u32, u32),
    User { name: String }
  }
  ```

  **Enum in TypeScript with `@scale-codec/enum`:**

  ```ts
  import { Enumerate } from '@scale-codec/enum'

  type Message = Enumerate<{
    Ping: []
    Text: [string]
    Coords: [[number, number]]
    User: [{ name: string }]
  }>
  ```
  
  Once enum type is declared, it could be used to create its **variants**.

  **Creating variants in Rust:**

  ```rust
  let mut msg = Message::Ping;
  msg = Message::Text("foo".to_owned());
  msg = Message::Coords(4, 2);
  msg = Message::User { name: "bar".to_owned() };
  ```

  **Creating variants in TypeScript with `@scale-codec/enum`:**

  ```ts
  import { variant } from '@scale-codec/enum'

  let msg: Message = variant('Ping')
  msg = variant('Text', 'foo')
  msg = variant('Coords', [4, 2])
  msg = variant('User', { name: 'bar' })
  ```

>   **Variant instance** has `tag` and `content` fields to work with:

Reviewer: ```suggestion
  Each **variant instance** has `tag` and `content` fields to work with:
```

  ```ts
  declare const msg: MessageVariant

  if (msg.tag === 'Text') {
    console.log(msg.content.trim())
  }
  ```
  

>   There are other changes as well, but that's all for major ones.

Reviewer: ```suggestion
  There are other changes as well, but that's all for the major ones.
```

  #### Why the change was made

>   The goal is to be consistent with [TypeScript support of discriminated unions](https://www.typescriptlang.org/docs/handbook/typescript-in-5-minutes-func.html#discriminated-unions). The consequence is that TypeScript now able to narrow types and to check for exhaustiveness in code that works with enums. In other words, **new enums are type-stronger**.

Reviewer: ```suggestion
  The goal is to be consistent with [TypeScript support of discriminated unions](https://www.typescriptlang.org/docs/handbook/typescript-in-5-minutes-func.html#discriminated-unions). As a consequence, TypeScript can now narrow types and check for exhaustiveness in code that works with enums. In other words, **new enums are type-stronger**.
```
